openapi_python_client/parser/properties/union.py

```diff
@@ -189,3 +199,105 @@ def validate_location(self, location: oai.ParameterLocation) -> ParseError | Non
             if evolve(cast(Property, inner_prop), required=self.required).validate_location(location) is not None:
                 return ParseError(detail=f"{self.get_type_string()} is not allowed in {location}")
         return None
+
+
+def _flatten_union_properties(
+    sub_properties: list[PropertyProtocol],
+) -> tuple[list[PropertyProtocol], list[DiscriminatorDefinition]]:
+    flattened = []
+    discriminators = []
+    for sub_prop in sub_properties:
+        if isinstance(sub_prop, UnionProperty):
+            if sub_prop.discriminators:
+                discriminators.extend(sub_prop.discriminators)
+            new_props, new_discriminators = _flatten_union_properties(sub_prop.inner_properties)
+            flattened.extend(new_props)
+            discriminators.extend(new_discriminators)
+        else:
+            flattened.append(sub_prop)
+    return flattened, discriminators
+
+
+def _parse_discriminator(
+    data: oai.Discriminator,
+    subtypes: list[PropertyProtocol],
+    schemas: Schemas,
+) -> DiscriminatorDefinition | PropertyError:
+    from .model_property import ModelProperty
+
+    # See: https://spec.openapis.org/oas/v3.1.0.html#discriminator-object
+
+    def _find_top_level_model(matching_model: ModelProperty) -> ModelProperty | None:
+        # This is needed because, when we built the union list, $refs were changed into a copy of
+        # the type they referred to, without preserving the original name. We need to know that
+        # every type in the discriminator is a $ref to a top-level type and we need its name.
+        for prop in schemas.classes_by_reference.values():
```

Comment:
What's going on here:

The generator defines `schemas.classes_by_reference` as a map that keeps track of all _named_ schemas— that is, ones that are defined in `components.schemas` rather than being anonymous inline schemas inside something else. So, for this spec—

```yaml
components:
  schemas:
    ModelA:
      # ...
    ModelB:
      # ...
    UnionAB:
      oneOf:
      - $ref: "#/components/schemas/ModelA"
      - $ref: "#/components/schemas/ModelB"
```

—the map will end up looking like this:
```python
{
  "/components/schemas/ModelA": model_property_instance_for_model_a,
  "/components/schemas/ModelB": model_property_instance_for_model_b,
  "/components/schemas/UnionAB": union_property_instance_for_union_ab,
}
```

It uses that map when it's resolving the `$ref`s inside UnionAB. However, when it does that, the `ModelProperty` instances that are inside of the `UnionProperty` are _not_ the same instances as `model_property_instance_for_model_a`, etc. They are newly parsed ones made from a copy of the YAML data of the original declaration. The only way to find the original instance is what I'm doing here: check the `class_info` attribute, which describes the Python class (`ModelA` or whatever) that will be used for this schema in the generated code.